Implement additional usermigrations

lukasgraf opened the pull request:
This PR adds support for additional usermigrations using `ftw.usermigration`, and "officially" registers the already existing user migrations directly in `opengever.core`.

Additional user migrations implemented in this PR:

- **OGDS User references**
  Migrates references to user IDs in OGDS:
    - activity actors
    - notification userids
    - task principals
    -  task issuers
    - task responsibles

---

- **Plone Tasks**
  Migrates
    - Plone Task responsibles
    - Plone Task issuers
    - Plone Task responses (actors + changes)

---

- **Private Folders**
    - Folders themselves
    - Merges contents between old and new private folders if necessary

---

- **Checked out documents**
    - checked_out_by annotations + index
    - WebDAV Locks

---

- **Repository favorites**

---

In addition, it contains an optimization for the `DossierMigrator` so it doesn't unnecessarily reindex `responsible` for every dossier.

**TODO**:

- [x] Changelog
- [x] Better PR description
- [x] Testing on backport-branch

phgross commented on opengever/usermigration/ogds_references.py:
```diff
@@ -0,0 +1,157 @@
+                moved = self._migrate_sql_column(
+                    Notification.__table__, 'userid',
+                    old_userid, new_userid)
+                self.notification_userids_moved.extend(moved)
```
Watchers.actorid needs to migrated as well.

phgross commented on opengever/usermigration/ogds_references.py:
```diff
@@ -0,0 +1,157 @@
+                    Task.__table__, 'responsible',
+                    old_userid, new_userid)
+                self.task_responsibles_moved.extend(moved)
+
```
What about OGDS Dictstorage entries?

lukasgraf replied:
As discussed, already implemented in [`opengever/usermigration/dictstorage.py`](https://github.com/4teamwork/opengever.core/blob/master/opengever/usermigration/dictstorage.py)

phgross commented on opengever/usermigration/plone_tasks.py:
```diff
@@ -0,0 +1,146 @@
+FIELDS_TO_CHECK = ('responsible', 'issuer')
+
+
+class PloneTasksMigrator(object):
```
I would add a comment here, that all the automatic created local roles on tasks, task documents, and task containers are handled by another Migrator.

phgross commented on setup.py:
```diff
@@ -110,6 +110,7 @@
           'ftw.table>=1.18.0',
           'ftw.tika',
           'ftw.upgrade >= 1.18.0',
+          'ftw.usermigration',
```
Please add also a version pinning to the `versions.cfg`

lukasgraf commented:
Thanks for the review! Updated - addressed all the feedback, added repo favorites migration that was missing, and fixed some issues with migration of task responses (see #3419).
